cmds/webboot/webboot.go

```diff
@@ -66,6 +59,52 @@ func (d *DownloadByLink) Exec() error {
 	return nil
 }
 
+// Exec displays subdirectory or cached isos under the path directory
+func (g *DirGroup) Exec() error {
+	entries := []menu.Entry{}
+	readerInfos, err := ioutil.ReadDir(g.path)
+	if err != nil {
+		return err
+	}
+
+	// check the directory, if there is a subdirectory, add another DirGroup option
+	// if there is iso file, add a cachedISO option
+	for _, info := range readerInfos {
+		if info.IsDir() {
+			entries = append(entries, &DirGroup{
+				label:    info.Name(),
+				path:     filepath.Join(g.path, info.Name()),
+				uiEvents: g.uiEvents,
+			})
+		} else if filepath.Ext(info.Name()) == ".iso" {
+			iso := &CachedISO{
+				path:  filepath.Join(g.path, info.Name()),
+				label: info.Name(),
+			}
+			entries = append(entries, iso)
+		}
+	}
+	_, err = menu.DisplayMenu("Distros", "Choose an option", entries, g.uiEvents)
+	return err
+}
+
+// Exec displays the next level of bookmarks
+func (g *BookmarkGroup) Exec() error {
+	nextLevel := []menu.Entry{}
+	for _, e := range g.entries {
+		if b, ok := e.(*BookmarkGroup); ok {
+			b.uiEvents = g.uiEvents
+			nextLevel = append(nextLevel, b)
+		} else {
```

Comment:
I think there is some redundancy but I cannot simplify it. In every BookmarkGroup option, I have to re-traverse all entries and add uiEvents to them one by one. So the benefit of direct hardcode bookmarks is reduced, I still need to reprocess entries instead of using them directly. But besides that, I have no other idea to add uiEvents to the entries in the next level. @pallaud

Reply:
For the sake of making this PR a little less complicated, can we just have the bookmarks show up as a single list for now, and not have any hierarchy? In a follow up PR we can look into doing that but for now I think it'd be better to just see this work first.